File: Linked Lists/linked list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
    
    def append(self, val):
        new_node = Node(val)

        if not self.head:
            self.head = new_node
            return
        
        curr = self.head

        while curr.next:
            curr = curr.next
        curr.next = new_node

    def delete_at_beginning(self):
        self.head = self.head.next
    
    def delete_at_end(self):
        curr = self.head
        while curr.next.next:
            curr = curr.next
        curr.next = None

    def delete_nth_node(self, n):
        dummy = LinkedList()
        dummy.next = self.head
        left = dummy
        curr = self.head

        while n > 0:
            curr = curr.next
            n -= 1
        
        while curr:
            left = left.next
            curr = curr.next
        left.next = left.next.next

        self.head = dummy.next


    def print_list(self):
        curr = self.head

        while curr:
            print(curr.data, end=" -> ")
            curr = curr.next
        print('None')

lst = LinkedList()
arr = [1,2,3,6,8,7,9]
for i in arr:
    lst.append(i)

lst.print_list()
# lst.delete_at_beginning()
# lst.print_list()
# lst.delete_at_end()
# lst.print_list()
lst.delete_nth_node(4)
lst.print_list()


        

File: CPP/Linked Lists/linked list.cpp
#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
    int data;
    Node *next;

public:
    Node(int data1, Node *next1)
    {
        data = data1;
        next = next1;
    }

public:
    Node(int data1)
    {
        data = data1;
        next = nullptr;
    }
};

Node *convertArr2LL(vector<int> arr)
{
    Node *head = new Node(arr[0]);
    Node *curr = head;

    for (int i = 1; i < arr.size(); i++)
    {
        Node *temp = new Node(arr[i]);
        curr->next = temp;
        curr = temp;
    }
    return head;
}

int lengthOfLL(Node *head)
{
    Node *curr = head;
    int count = 0;

    while (curr)
    {
        count++;
        curr = curr->next;
    }

    return count;
}

void print(Node *head)
{
    Node *curr = head;

    while (curr)
    {
        cout << curr->data << " -> ";
        curr = curr->next;
    }
    cout << "null" << endl;
}

int main()
{
    vector<int> arr = {2, 4, 12, 8, 3};
    Node *head = convertArr2LL(arr);

    print(head);
    cout << lengthOfLL(head) << endl;

    return 0;
}